src/libraries/System.Private.CoreLib/src/System/Threading/Wasi/WasiEventLoop.cs

```diff
@@ -29,18 +32,71 @@ internal static Task RegisterWasiPollable(Pollable pollable, CancellationToken c
             // this will register the pollable holder into s_pollables
             var holder = new PollableHolder(pollable, cancellationToken);
             s_pollables.Add(holder);
+
+            ScheduleCheck();
+
             return holder.taskCompletionSource.Task;
         }
 
-        // this is not thread safe
-        internal static void DispatchWasiEventLoop()
+
+        internal static T PollWasiEventLoopUntilResolved<T>(Task<T> mainTask)
+        {
+            s_mainTask = mainTask;
+            while (!mainTask.IsCompleted)
+            {
+                ThreadPoolWorkQueue.Dispatch();
+            }
+            s_mainTask = null;
```

Comment:
Should we consider putting this in a `finally` block?  In case `ThreadPoolWorkQueue.Dispatch` throws an exception?  Or is that not possible?